docs/design/2022-05-19-compact-table-via-sql.md

```diff
@@ -0,0 +1,224 @@
+  won’t return until the compaction is done.
+- When a table is already in compacting progress, the new compaction SQL command
+  involving this table will be exited immediately by producing a “canceled”
+  warning.
+- The compact SQL commands can be executed on different tables simultaneously,
+  resulting in multiple tables being compacted simultaneously.
+- When `engine_type` is specified as `TIKV`, “unsupported” error will be
+  returned. When `engine_type` is not specified, the SQL will run as compact
+  TiFlash replicas only. This behavior will change when we support TiKV compaction
+  in future.
+- When the table contains multiple partitions, all partitions will be compacted.
+- The compact command will exit in the following ways:
+  1. User kill SQL via `KILL [TIDB] <ID>`: stop task immediately
+  2. TiDB stopped: the compaction run on TiFlash should be stopped. There will
+     be no retries after TiDB is restarted.
+  3. TiFlash stopped: the compaction command should stop and return error
+     immediately. There will be no retries after TiFlash is restarted.
+  4. Compaction is finished.
+
+## Detailed Design
+
+### Protocol
+
+New APIs will be added to TiFlash:
+
+```protobuf
+// Pseudo code
+
+message Error {
+    oneof {
+        ErrorCompactInProgress,
+        ErrorTooManyPendingTasks,
+        // More errors can be added in future
+    }
+}
+
+message ErrorCompactInProgress {}
+
+message ErrorTooManyPendingTasks {}
+
+message CompactRequest {
+    bytes id
+    bytes start_key   // Optional
+    bytes max_end_key // Optional
+    int64 physical_table_id // Used to locate the TiFlsh table
+}
+
+message CompactResponse {
+    optional Error error
+    bool has_remaining
+    bytes compacted_start_key
+    bytes compacted_end_key
+}
+```
+
+### General Flow
+
+![](./images/2022-05-19-compact-table-via-sql-1.png)
+
+TiDB sends `CompactRequest` to one TiFlash instance in series. Each request
+compacts one or multiple Segments in the TiFlash instance. The segments may
+change (e.g. split or merge) during the SQL execution process so that TiFlash
+will respond with the end key for each request, which will be then used as the
+`StartKey` for the next request sent from TiDB.
+
+When there are multiple TiFlash instances, TiDB talks with each TiFlash
+concurrently. However, for each TiFlash connection, requests are still sent in
+series. Newly added TiFlash instances during a compaction SQL execution are
+discarded.
+
+### Interrupt Execution (Kill)
+
+![](./images/2022-05-19-compact-table-via-sql-2.png)
+
+When user executes `KILL [TIDB] <ID>` to interrupt the compaction command,
+TiDB simply stops the execution by not sending future `CompactRequest`.
+There will be `CompactRequest` running in TiFlash instances at the moment when
+user initiates the `KILL`. These running "small" compaction tasks will be
+untouched and kept running until finished. There is no way to stop a running
```

Comment:
What will happen to the intermediate files if a compact task is stopped?

Reply:
TiFlash only accepts segment level compaction requests and these requests cannot be stopped once started. This means, the compaction for one segment will be finished. TiDB will not send more segment level compaction requests, causing the resource usage to drop after KILLing the TiDB SQL execution.